jstest/testrunner/devices/device_base.py

```diff
@@ -29,8 +32,23 @@ def __init__(self, env):
         self.device = env['info']['device']
         self.env = env
 
+
         self.check_args()
 
+    @abstractproperty
+    def workdir(self):
```

Comment:
I wouldn't use property annotations. You can put these properties into the `__init__` function with `None` value. The concrete subclasses can override their values.